### PR TITLE
docs(NODE-3610): update contributing guidelines

### PR DESCRIPTION
**Pull request by dariakp**
## Description

Cleaned up the existing contributing guide and added reviewer guidelines.

### CODE REVIEW

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -116,11 +117,42 @@ Below are some conventions that aren't enforced by any of our tooling but we non
...
+     - If the change is breaking, ensure there is an exclamation mark after the scope (e.g., "fix(NODE-xxx)!: \<description\>" )
+  1. If there are new TODOs, has a related JIRA ticket been created?
+  1. Are symbols correctly marked as internal or public?
+  1. Do the Typescript types match expected runtime usage? Are there tests for new or updated types?
+  1. Should any documentation be updated?
+     - Has the relevant internal documentation been updated as part of the PR?
+     - Have the external documentation requirements been captured in jira?
+
+[conventional-commit-style]: https://www.conventionalcommits.org/en/v1.0.0/
 [changelog]: CHANGELOG.md
```
Comment (nbbeeken) [inline CONTRIBUTING.md:156]: I don't think we're using this, but it should point to HISTORY.md if we do

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -116,11 +117,42 @@ Below are some conventions that aren't enforced by any of our tooling but we non
...
+     - If there are associated automated spec tests, were they all pulled in and are they all running and correctly interpreting the spec inputs?
+       - Are any runner changes needed to process new input types?
+  1. Could these changes impact any adjacent functionality?
+  1. Are there any errors that might not be correctly caught or propagated?
+  1. Is there anything that could impact performance?
+  1. Are there any race conditions in the functional code or tests?
+  1. Can you think of a better way to implement any of the functional code or tests? "Better" means any combination of:
+     - more performant
+     - better organized / easier to understand
+     - easier to maintain (easier to change, harder to accidentally break)
```
Comment (nbbeeken) [inline CONTRIBUTING.md:144]: an additional point to the maintain parenthetical: `(are modular as in makes good use of before/after hooks to keep actual testing code clear)`

wordy but I think worth capturing, thoughts?

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -116,11 +117,42 @@ Below are some conventions that aren't enforced by any of our tooling but we non
...
+  1. Does the code meet the acceptance criteria?
+     - If there is an associated spec, does the code match the spec?
+  1. Is the intention of the code captured in relevant tests?
+     - Does the description of each test accurately represent the assertions?
+     - For any test explicitly called out on the ticket as desirable to implement, was it implemented?
+     - If there are prose spec tests, were they implemented?
+     - If there are associated automated spec tests, were they all pulled in and are they all running and correctly interpreting the spec inputs?
+       - Are any runner changes needed to process new input types?
+  1. Could these changes impact any adjacent functionality?
+  1. Are there any errors that might not be correctly caught or propagated?
```
Comment (nbbeeken) [inline CONTRIBUTING.md:138]: Stop me if I'm getting to implementation specific

A sub-bullet here about mocha specifically, testing event handling and how error throwing can make unrelated tests fail.

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -116,18 +117,47 @@ Below are some conventions that aren't enforced by any of our tooling but we non
...
+  1. Is there anything that could impact performance?
+  1. Are there any race conditions in the functional code or tests?
+  1. Can you think of a better way to implement any of the functional code or tests? "Better" means any combination of:
+     - more performant
+     - better organized / easier to understand
```
Comment (dariakp) [inline CONTRIBUTING.md:143]:
```suggestion
     - better organized / easier to understand / clearer separation of concerns
```

**Pull request merged**